### DISCUSSION TITLE
Does rust analyzer supports expanding proc-macros that read files?

### DISCUSSION THREAD

**Discussion by Davidblkx** [rust-lang/rust-analyzer#16916] [Q&A] [answered]
I was playing arround with proc-marcos and tried todo a simple one:

```rust
#![feature(proc_macro_span)]
use proc_macro::{Span, TokenStream};

extern crate proc_macro;

#[proc_macro]
pub fn declare_module(raw: TokenStream) -> TokenStream {
    let mut path = Span::call_site().source_file().path();
    let input = syn::parse_macro_input!(raw as syn::Ident);

    path.pop();
    path.push(format!("{}.rs", input.to_string()));

    let text_file = std::fs::read_to_string(path).unwrap();
    return  text_file.parse().unwrap();
}
``` 

If I try to run it, it compiles and runs ok but during development r-a doesn't expand the macro, if I replace this logic by just `"..my code...".parse().unwrap()` everything works ok and I have auto-complete inside VSCode.

I also run the vscode command: `Expand macro recursively at caret` and it just opens a empty file

I can't find any error log for r-a in vscode, so I'm wondering if I can do something to make it work or if it is at all supported

**Comment by Veykril** [comment, marked as answer]
Most of the `proc_macro_span` feature is currently unimplemented in r-a

**Comment by Davidblkx** [comment]
That explains it, thanks 😄
